stdlib/mqtt/to.go

```diff
@@ -0,0 +1,649 @@
+func (ToMQTTOpSpec) Kind() flux.OperationKind {
+	return ToMQTTKind
+}
+
+type ToMQTTProcedureSpec struct {
+	plan.DefaultCost
+	Spec *ToMQTTOpSpec
+}
+
+func (o *ToMQTTProcedureSpec) Kind() plan.ProcedureKind {
+	return ToMQTTKind
+}
+
+func (o *ToMQTTProcedureSpec) Copy() plan.ProcedureSpec {
+	s := o.Spec
+	res := &ToMQTTProcedureSpec{
+		Spec: &ToMQTTOpSpec{
+			Broker:          s.Broker,
+			Topic: 		  s.Topic,
+			Name:         s.Name,
+			QoS:          s.QoS,
+			Username:     s.Username,
+			Password:     s.Password,
+			NameColumn:   s.NameColumn,
+			Timeout:      s.Timeout,
+			NoKeepAlive:  s.NoKeepAlive,
+			TimeColumn:   s.TimeColumn,
+			TagColumns:   append([]string(nil), s.TagColumns...),
+			ValueColumns: append([]string(nil), s.ValueColumns...),
+		},
+	}
+	return res
+}
+
+func newToMQTTProcedure(qs flux.OperationSpec, a plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*ToMQTTOpSpec)
+	if !ok && spec != nil {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &ToMQTTProcedureSpec{Spec: spec}, nil
+}
+
+func createToMQTTTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*ToMQTTProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewToMQTTTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type ToMQTTTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+	spec  *ToMQTTProcedureSpec
+}
+
+func (t *ToMQTTTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func NewToMQTTTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *ToMQTTProcedureSpec) *ToMQTTTransformation {
+	return &ToMQTTTransformation{
+		d:     d,
+		cache: cache,
+		spec:  spec,
+	}
+}
+
+type toMqttMetric struct {
+	tags   []*protocol.Tag
+	fields []*protocol.Field
+	name   string
+	t      time.Time
+}
+
+func (m *toMqttMetric) TagList() []*protocol.Tag {
+	return m.tags
+}
+func (m *toMqttMetric) FieldList() []*protocol.Field {
+	return m.fields
+}
+
+func (m *toMqttMetric) truncateTagsAndFields() {
+	m.fields = m.fields[:0]
+	m.tags = m.tags[:0]
+}
+
+func (m *toMqttMetric) Name() string {
+	return m.name
+}
+
+func (m *toMqttMetric) Time() time.Time {
+	return m.t
+}
+
+type idxType struct {
+	Idx  int
+	Type flux.ColType
+}
+
+func (t *ToMQTTTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	// set up the MQTT options.
+	opts := MQTT.NewClientOptions().AddBroker(t.spec.Spec.Broker)
+	if len(t.spec.Spec.ClientID) > 0 {
+		opts.SetClientID(t.spec.Spec.ClientID)
+	} else {
+		opts.SetClientID("flux-mqtt")
+	}
+	if t.spec.Spec.Timeout > 0 {
+		opts.SetConnectTimeout(t.spec.Spec.Timeout)
+	}
+	if len(t.spec.Spec.Username) > 0 {
```

Comment:
if these are strings, it might be clearer to say `if t.spec.Spec.Username != ""`

Reply:
Done!